pkg/services/ngalert/schedule/registry.go

```diff
@@ -0,0 +1,156 @@
+package schedule
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/services/alerting"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+type alertRuleRegistry struct {
+	mu            sync.Mutex
+	alertRuleInfo map[models.AlertRuleKey]*alertRuleInfo
+}
+
+// getOrCreateInfo gets rule routine information from registry by the key. If it does not exist, it creates a new one.
+// Returns a pointer to the rule routine information and a flag that indicates whether it is a new struct or not.
+func (r *alertRuleRegistry) getOrCreateInfo(context context.Context, key models.AlertRuleKey) (*alertRuleInfo, bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	info, ok := r.alertRuleInfo[key]
+	if !ok {
+		info = newAlertRuleInfo(context)
+		r.alertRuleInfo[key] = info
+	}
+	return info, !ok
+}
+
+// get returns the channel for the specific alert rule
+// if the key does not exist returns an error
+func (r *alertRuleRegistry) get(key models.AlertRuleKey) (*alertRuleInfo, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	info, ok := r.alertRuleInfo[key]
+	if !ok {
+		return nil, fmt.Errorf("%v key not found", key)
+	}
+	return info, nil
+}
+
+func (r *alertRuleRegistry) exists(key models.AlertRuleKey) bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	_, ok := r.alertRuleInfo[key]
+	return ok
+}
+
+// del removes pair that has specific key from alertRuleInfo.
+// Returns 2-tuple where the first element is value of the removed pair
+// and the second element indicates whether element with the specified key existed.
+func (r *alertRuleRegistry) del(key models.AlertRuleKey) (*alertRuleInfo, bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	info, ok := r.alertRuleInfo[key]
+	if ok {
+		delete(r.alertRuleInfo, key)
+	}
+	return info, ok
+}
+
+func (r *alertRuleRegistry) iter() <-chan models.AlertRuleKey {
+	c := make(chan models.AlertRuleKey)
+
+	f := func() {
+		r.mu.Lock()
+		defer r.mu.Unlock()
```

Comment:
I think this can deadlock if the consumer doesn't read the channel (i.e. it receives cancelation). It will also block all other goroutines if the consumer takes a long time to iterate all of the items.

Reply:
I copied the code as-is from the scehdule.go.

Reply:
Then let's file a bug somewhere, too.

Reply:
> I think this can deadlock if the consumer doesn't read the channel (i.e. it receives cancelation). It will also block all other goroutines if the consumer takes a long time to iterate all of the items.

I took a look at this closer, and although I do not understand why such a simple operation needs to be that complex, I do not see a case when the situation can happen. The method is used only in one place, a few lines below, the channel is read until it's closed. I would appreciate some examples of when it can fail.